### PR TITLE
Refactor DAG for easier KFS integration

### PR DESCRIPTION
**Pull request by atobiszei**
Additionally part of preparation for single model prediction path integration.

JIRA:CVS-81054, CVS-81052

### CODE REVIEW

**Review by mkuczyns11 — APPROVED** [review]

File: src/predict_request_validation_utils.cpp [hunk 1]
```diff
@@ -367,9 +368,15 @@ Status RequestValidator::validate() {
     return finalStatus;
 }
 
+template <>
 Status validate(const tensorflow::serving::PredictRequest& request, const tensor_map_t& inputsInfo, const std::string& servableName, const model_version_t servableVersion, const std::set<const char*>& optionalAllowedInputNames, const Mode batchingMode, const shapes_info_map_t& shapeInfo) {
     return RequestValidator(request, inputsInfo, servableName, servableVersion, optionalAllowedInputNames, batchingMode, shapeInfo).validate();
 }
 
+template <>
+Status validate(const ::inference::ModelInferRequest& request, const tensor_map_t& inputsInfo, const std::string& servableName, const model_version_t servableVersion, const std::set<const char*>& optionalAllowedInputNames, const Mode batchingMode, const shapes_info_map_t& shapeInfo) {
```
Comment (dkalinowski) [inline src/predict_request_validation_utils.cpp:377]: If KFS has no namespace. maybe we could add it? kfs::inference::ModelInferRequest

File: src/predict_request_validation_utils.hpp [hunk 2]
```diff
@@ -15,6 +15,7 @@
 //*****************************************************************************
 #pragma once
 
+// TODO move into cpp
```
Comment (dkalinowski) [inline src/predict_request_validation_utils.hpp:18]: Why not now?

File: src/serialization.hpp [hunk 3]
```diff
@@ -43,29 +44,47 @@ class OutputGetter {
     T outputSource;
 };
 
+template <typename ProtoStorage, typename ProtoType>
+class ProtoGetter {
+    ProtoStorage protoStorage;
+
+public:
+    ProtoGetter(ProtoStorage protoStorage) :
```
Comment (dkalinowski) [inline src/serialization.hpp:52]: Wouldn't it be passed by copy?

File: src/test/ensemble_flow_custom_node_tests.cpp [hunk 4]
```diff
@@ -386,17 +386,17 @@ TEST_F(EnsembleFlowCustomNodeAndDemultiplexerGatherPipelineExecutionTest, Multip
...
+    nodes[nodesCount - 1] = std::make_unique<ExitNode<PredictResponse>>(&response, outputsInfo);
     size_t i = 1;
     for (size_t demultiplicationLayer = 0; demultiplicationLayer < demultiplicationLayersCount; ++demultiplicationLayer) {
         nodes[i++] = std::make_unique<CustomNode>(differentOpsNodeName + "-" + std::to_string(demultiplicationLayer), differentOpsLibrary, parameters_t{}, differentOpsOutputAlias, demultiplyCount);
         nodes[i++] = std::make_unique<DLNode>(dummyNodeName + "-" + std::to_string(demultiplicationLayer), "dummy", std::nullopt, modelManager);
         nodes[nodesCount - 1 - (i / 2)] = std::make_unique<CustomNode>(chooseMaxNodeName + "-" + std::to_string(demultiplicationLayer), chooseMaxLibrary, parameters, chooseMaxOutputAlias, std::nullopt, std::set<std::string>({differentOpsNodeName + "-" + std::to_string(demultiplicationLayer)}));
     }
 
-    Pipeline pipeline(dynamic_cast<EntryNode&>(*nodes[0]), dynamic_cast<ExitNode&>(*nodes[nodesCount - 1]));
+    Pipeline pipeline(dynamic_cast<EntryNode<PredictRequest>&>(*nodes[0]), dynamic_cast<ExitNode<PredictResponse>&>(*nodes[nodesCount - 1]));
```
Comment (dkalinowski) [inline src/test/ensemble_flow_custom_node_tests.cpp:399]: Why do you need to cast here? Pipeline accepts type "Node"

Comment (atobiszei) [reply src/test/ensemble_flow_custom_node_tests.cpp:399]: Good catch

File: src/serialization.hpp (hunk 3 above)
Comment (atobiszei) [reply src/serialization.hpp:52]: It depends on what type you create -
if
ProtoGetter<tensorflow::serving::PredictResponse*, tensorflow::TensorProto&> then no
if
ProtoGetter<tensorflow::serving::PredictResponse, tensorflow::TensorProto> then yes

File: src/predict_request_validation_utils.cpp (hunk 1 above)
Comment (atobiszei) [reply src/predict_request_validation_utils.cpp:377]: kfs has ::inference namespace

File: src/modelinstance.hpp [hunk 5]
```diff
@@ -218,7 +218,8 @@ class ModelInstance {
          */
     Status loadOVModelUsingCustomLoader();
 
-    virtual const Status validate(const tensorflow::serving::PredictRequest* request);
+    template <typename PredictRequest>
```
Comment (ficol) [inline src/modelinstance.hpp:221]: Maybe change ```<typename PredictRequest>``` to ```<typename RequestType>``` to be consistent.

**Review by ficol — APPROVED** [review]

**Review by dkalinowski — APPROVED** [review]

**Pull request merged**